File: tests/OpenTK.Tests.CSharp/Quaternion_Tests.cs
using NUnit.Framework;
using System;
using System.Collections;
namespace OpenTK.Tests.CSharp
{
	/// <summary>
	/// Quaternion test data.
	/// </summary>
	internal class QuaternionTestData
	{
		/// <summary>
		/// Gets the single axis test cases.
		/// </summary>
		/// <value>The single axis test cases.</value>
		public static IEnumerable SingleAxisTestCases
		{
			get
			{
				yield return new TestCaseData(new Vector3(1, 0, 0), Vector3.UnitX).SetName("Rotate around x axis");
				yield return new TestCaseData(new Vector3(0, 1, 0), Vector3.UnitY).SetName("Rotate around y axis");
				yield return new TestCaseData(new Vector3(0, 0, 1), Vector3.UnitZ).SetName("Rotate around z axis");
			}
		}

		public static IEnumerable ToAxisAngleTestCases
		{
			get
			{
				yield return new TestCaseData(new Quaternion(Vector3.UnitX, 0), Vector3.UnitX).SetName("Rotate around x axis");
				yield return new TestCaseData(new Quaternion(Vector3.UnitY, 0), Vector3.UnitY).SetName("Rotate around y axis");
				yield return new TestCaseData(new Quaternion(Vector3.UnitZ, 0), Vector3.UnitZ).SetName("Rotate around z axis");
			}
		}
	}

	/// <summary>
	/// Provides some methods which helps to verify test results
	/// </summary>
	internal static class QuaternionTestHelper
	{
		/// <summary>
		/// Verifies the direction of an given <see cref="Vector3"/>
		/// </summary>
		/// <returns>The direction.</returns>
		/// <param name="toTest">To test.</param>
		/// <param name="expected">Expected.</param>
		public static bool VerifyEuqalDirection(Vector3 toTest, Vector3 expected)
		{
			//To verify the direction of an vector, just respect the 0 values and check against these.
			//The length of the vectors are ignored.
			if (expected.X == 0)
			{
				if (toTest.X != 0)
					return false;
			}

			if (expected.Y == 0)
			{
				if (toTest.Y != 0)
					return false;
			}

			if (expected.Z == 0)
			{
				if (toTest.Z != 0)
					return false;
			}

			return true;
		}
	}

	[TestFixture()]
	public class Quaternion_Tests
	{
		/// <summary>
		/// Checks if a single given value (either pitch, yaw or roll) get converted into correct x,y,z value of quaternion.
		/// </summary>
		/// <param name="rotationAxis">euler angle values</param>
		[Test, TestCaseSource(typeof(QuaternionTestData), "SingleAxisTestCases")]
		public void CtorEulerAngles_SingleEulerAngleSet_RotateCorrectAxis(Vector3 eulerValues, Vector3 expectedResult)
		{
			//Arrange + Act: Create Quaternion with "pitch/yaw/roll"
			var cut = new Quaternion(eulerValues.X, eulerValues.Y, eulerValues.Z);

			//Assert: Use helper, to check if part of the two correct axis is zero. I just want check the direction
			Vector3 resultXYZ = cut.Xyz;

			Assert.IsTrue(QuaternionTestHelper.VerifyEuqalDirection(resultXYZ, expectedResult));
		}

		/// <summary>
		/// Checks if a single given value (either pitch, yaw or roll) get converted into correct x,y,z value of quaternion.
		/// </summary>
		/// <param name="rotationAxis">euler angle values</param>
		[Test, TestCaseSource(typeof(QuaternionTestData), "SingleAxisTestCases")]
		public void FromEulerAngles_SingleEulerAngleSet_RotateCorrectAxis(Vector3 eulerValues, Vector3 expectedResult)
		{
			//Arrange + Act: Create Quaternion with "pitch/yaw/roll"
			var cut = Quaternion.FromEulerAngles(eulerValues.X, eulerValues.Y, eulerValues.Z);

			//Assert: Use helper, to check if part of the two correct axis is zero. I just want check the direction
			Vector3 resultXYZ = cut.Xyz;

			Assert.IsTrue(QuaternionTestHelper.VerifyEuqalDirection(resultXYZ, expectedResult));
		}

		/// <summary>
		/// Check if a quaternion returns a a rotation about the correct coordinate axis
		/// </summary>
		/// <returns>The axis angle single axis set and angle ignored rotate correct axis.</returns>
		/// <param name="cut">Cut.</param>
		/// <param name="expectedResult">Expected result.</param>
		[Test, TestCaseSource(typeof(QuaternionTestData), "ToAxisAngleTestCases")]
		public void ToAxisAngle_SingleAxisSetAndAngleIgnored_RotateCorrectAxis(Quaternion cut, Vector3 expectedResult)
		{
			//Arrange + Act: Create Quaternion with rotation about X/Y/Z axis
			Vector3 resultXYZ;
			float dontCare;
			cut.ToAxisAngle(out resultXYZ, out dontCare);

			//Assert: Use helper, to check if part of the two correct axis is zero. I just want check the direction
			Assert.IsTrue(QuaternionTestHelper.VerifyEuqalDirection(resultXYZ, expectedResult));
		}

		//TODO: Make also checks with rotation angle
	}
}

File: tests/OpenToolkit.Tests/QuaternionTests.fs
namespace OpenToolkit.Tests

open Xunit
open FsCheck.Xunit
open OpenToolkit.Mathematics

[<Properties(Arbitrary = [| typeof<OpenTKGen> |])>]
module Quaternion =
    [<AutoOpen>]
    module Helpers =
         //To verify the direction of an vector, just respect the 0 values and check against these.
         //The length of the vectors are ignored.
        let verifyDirection(expected:Vector3) (toTest:Vector3) =
            Assert.Equal(sign expected.X, sign toTest.X)
            Assert.Equal(sign expected.Y, sign toTest.Y)
            Assert.Equal(sign expected.Z, sign toTest.Z)
            
        let axes = [
            Vector3.UnitX
            Vector3.UnitY
            Vector3.UnitZ
        ]
    
    [<Fact>]
    let ``Single axis as euler angles is converted to correct quaternion components``() =
        let test(v:Vector3) =
            let cut = Quaternion(v.X, v.Y, v.Z)
            verifyDirection cut.Xyz v
            
        axes |> List.iter test
        
    [<Fact>]
    let ``Single axis euler angles in vector3 is converted to correct quaternion components``() =
        let test(v:Vector3) =
            let cut = Quaternion(v)
            verifyDirection cut.Xyz v
            
        axes |> List.iter test
    
    [<Property>]
    let ``Quaternion euler angle construction methods produce consistent results``(v:Vector3) =
        let q1 = Quaternion(v)
        let q2 = Quaternion(v.X, v.Y, v.Z)
        let q3 = Quaternion.FromEulerAngles(v)
        let q4 = Quaternion.FromEulerAngles(v.X, v.Y, v.Z)
        let mutable q5 = Quaternion.Identity
        let mutable inV = v
        Quaternion.FromEulerAngles(&inV, &q5)
        
        let cases = [q2; q3; q4; q5]
        for case in cases do
             Assert.Equal(q1, case)
        
    [<Fact>]
    let ``Single rotation axis quaternion converts to correct axis angle representation``() =
        let test(v:Vector3) =
            let q = Quaternion(v)
            let mutable vec = Vector3()
            let mutable dontCare = 0.0f
            q.ToAxisAngle(&vec, &dontCare)
            verifyDirection v vec
            
        axes |> List.iter test
